### PR TITLE
template weighting update with ctf, dose weighting, and spectrum whitening

### PR DESCRIPTION
**Pull request by McHaillet**
This a quite a big PR that adds a lot of optional template weighting functionality based on the '3D-CTF' model. The template is weighted per tilt by 3 things:

- a simple CTF function determined by a per tilt estimated defocus value, voltage, amplitude contrast, and spherical aberration.
- a dose dependent Gaussian fall-off, which is calculated from the B-factor. 
- a tilt angle weighting taken as the cosine of the tilt angle, this models signal dampening due to increased sample thickness.

Secondly I added a whitening filter. This is a filter that flattens the power spectrum of the tomogram to a tomogram with white noise. It downweighs low frequencies and upweighs higher spatial frequencies. It sharpens the correlation peaks in the final score volume.

I made the following API changes in src/pytom_tm/weights.py :
- _create_structured_wedge() is renamed to _create_tilt_weighted_wedge() for clearity about functionality
- in create_wedge() the nyquist cut_off_radius parameter is now a keyword as all of the code uses a default value for it
- in create_wedge() the voxel_size parameter is now a positional argument, it is crucial to the wedge creation functionality (i.e. for calculating filters), so its clearer for future development when its not a keyword.
- version is changed to 0.3.0 in accordance with these API changes

@sroet, I know its quite some changes, so please let me know if you require some more detailed explanations!

### CODE REVIEW

File: src/pytom_tm/weights.py [hunk 1]
```diff
@@ -53,17 +54,25 @@ def wavelength_ev2m(voltage: float) -> float:
...
-        -shape[0] // 2 + shape[0] % 2,
-        shape[0] // 2 + shape[0] % 2, 1.
-    )) / (shape[0] // 2))[:, np.newaxis, np.newaxis]
-    y = (np.abs(np.arange(
-        -shape[1] // 2 + shape[1] % 2,
-        shape[1] // 2 + shape[1] % 2, 1.
-    )) / (shape[1] // 2))[:, np.newaxis]
-    z = np.arange(0, shape[2] // 2 + 1, 1.) / (shape[2] // 2)
-    return np.sqrt(x ** 2 + y ** 2 + z ** 2)
+def radial_reduced_grid(shape: Union[tuple[int, int, int], tuple[int, int]]) -> npt.NDArray[float]:
```
Comment (sroet) [inline src/pytom_tm/weights.py:57]: this returns `None` if `shape not in [2,3]` is that intended? If not, please catch and throw and error.
Either way, please add a test for the behavior

File: src/bin/pytom_match_template.py [hunk 2]
```diff
@@ -55,13 +57,44 @@ def main():
...
     parser.add_argument('-g', '--gpu-ids', nargs='+', type=int, required=True,
                         help='GPU indices to run the program on.')
     parser.add_argument('--log', type=str, required=False, default=20, action=ParseLogging,
                         help='Can be set to `info` or `debug`')
     args = parser.parse_args()
     logging.basicConfig(level=args.log)
 
+    # combine ctf values to ctf_params list of dicts
+    ctf_params = None
+    if isinstance(args.defocus_file, list):
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/bin/pytom_match_template.py:87]:
```suggestion
    if args.defocus_file is not None:
```

File: src/pytom_tm/io.py [hunk 3]
```diff
@@ -123,5 +149,28 @@ def read_mrc(
...
+        lines = fstream.readlines()
+    if file_name.suffix == '.defocus':
+        imod_defocus_version = float(lines[0].strip().split()[5])
+        # imod defocus files have the values specified in nm: TODO is this the common way to specify it?
+        if imod_defocus_version == 2:  # file with one defocus value and starts on line 0
+            return [float(x.strip().split()[4]) * 1e-3 for x in lines]
+        elif imod_defocus_version == 3:  # file with astigmatism and first line empty
+            return [(float(x.strip().split()[4]) + float(x.strip().split()[5])) / 2 * 1e-3 for x in lines[1:]]
+        else:
+            raise ValueError('Invalid IMOD defocus file inversion, can only be 2 or 3.')
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/io.py:172]:
```suggestion
    if file_name.suffix == '.defocus':
        imod_defocus_version = float(lines[0].strip().split()[5])
        # imod defocus files have the values specified in nm: TODO is this the common way to specify it?
        if imod_defocus_version == 2:  # file with one defocus value and starts on line 0
            return [float(x.strip().split()[4]) * 1e-3 for x in lines]
        elif imod_defocus_version == 3:  # file with astigmatism, first line is skipped as it contains only metadata
```

[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/io.py:172]:
```suggestion
    if file_name.suffix == '.defocus':
        imod_defocus_version = float(lines[0].strip().split()[5])
        # imod defocus files have the values specified in nm: TODO is this the common way to specify it?
        if imod_defocus_version == 2:  # file with one defocus value; data starts on line 0
            return [float(x.strip().split()[4]) * 1e-3 for x in lines]
        elif imod_defocus_version == 3:  # file with astigmatism; line 0 contains metadata that we do not need
            return [(float(x.strip().split()[4]) + float(x.strip().split()[5])) / 2 * 1e-3 for x in lines[1:]]
        else:
            raise ValueError('Invalid IMOD defocus file inversion, can only be 2 or 3.')
```

File: src/pytom_tm/tmjob.py [hunk 4]
```diff
@@ -33,7 +33,10 @@ def load_json_to_tmjob(file_name: pathlib.Path) -> TMJob:
         search_z=data['search_z'],
         voxel_size=data['voxel_size'],
         low_pass=data['low_pass'],
-        high_pass=data['high_pass']
+        high_pass=data['high_pass'],
+        dose_accumulation=data['dose_accumulation'],
+        ctf_data=data['ctf_data'],
+        whiten_spectrum=data['whiten_spectrum'],
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/tmjob.py:39]: thanks for spotting that

File: src/pytom_tm/weights.py [hunk 5]
```diff
@@ -142,12 +151,14 @@ def create_gaussian_band_pass(
 def create_wedge(
         shape: tuple[int, int, int],
         tilt_angles: list[float, ...],
-        cut_off_radius: float,
+        voxel_size: float,
+        cut_off_radius: float = 1.,
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/weights.py:157]: changed the version to 0.3.0 and also updated the PR description

File: src/pytom_tm/weights.py [hunk 6]
```diff
@@ -453,3 +538,69 @@ def radial_average(image: npt.NDArray[float]) -> tuple[npt.NDArray[float], npt.N
     mean = np.vectorize(lambda x: np.fft.fftshift(image, axes=(0, 1))[(r >= x - .5) & (r < x + .5)].mean())(q)
 
     return q, mean
+
+
+def power_spectrum_profile(image: npt.NDArray[float]) -> npt.NDArray[float]:
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/weights.py:552]: added!

File: src/pytom_tm/weights.py [hunk 7]
```diff
@@ -453,3 +538,69 @@ def radial_average(image: npt.NDArray[float]) -> tuple[npt.NDArray[float], npt.N
...
+
+    power_profile = np.vectorize(
+        lambda x: np.fft.fftshift(power, axes=(0, 1) if len(image.shape) == 3 else 0)[
+            (q_grid >= x - .5 * q_step) & (q_grid < x + .5 * q_step)
+        ].mean())(q)
+
+    return power_profile
+
+
+def profile_to_weighting(
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/weights.py:582]: added!

File: src/pytom_tm/weights.py [hunk 8]
```diff
@@ -310,40 +348,87 @@ def _create_structured_wedge(
...
+                    rotated *
+                    np.cos(alpha) *  # apply tilt-dependent weighting
+                    np.exp(-2 * np.pi ** 2 * sigma_motion ** 2 * q_squared)  # apply dose-weighting
+            )
+        else:
+            weighted_tilt = (
+                    rotated *
+                    np.cos(alpha)  # apply tilt-dependent weighting
+            )
+        tilt_sum[weighted_tilt > tilt_sum] = weighted_tilt[weighted_tilt > tilt_sum]
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/weights.py]: switched to np.maximum!

**Comment by McHaillet** [comment]
I updated/answered all the suggestions. Let me know if you have any other points ;)

**Review by sroet — APPROVED** [review]
Thanks for the fixes, LGTM feel free to merge

**Comment by McHaillet** [comment]
Thanks for the review!

**Pull request merged**